database/bfgd/postgres/postgres.go

```diff
@@ -965,6 +965,62 @@ func (p *pgdb) AccessPublicKeyDelete(ctx context.Context, publicKey *bfgd.Access
 	return nil
 }
 
+// BtcBlocksHeightsWithNoChildren returns the heights of blocks that have no
+// child blocks in our database, these represent possible forks that we have
+// not handled yet
+func (p *pgdb) BtcBlocksHeightsWithNoChildren(ctx context.Context) ([]uint64, error) {
+	log.Tracef("BtcBlocksHeightsWithNoChildren")
+	defer log.Tracef("BtcBlocksHeightsWithNoChildren exit")
+
+	/*
+		give me all heights for btc blocks
+		where there does not exist a child of this block
+		and there are no other blocks at this height with children
+		exclude the tip because it won't have children by definition
+	*/
+	const q = `
+		SELECT height FROM btc_blocks bb1
+
+		WHERE NOT EXISTS (SELECT * FROM btc_blocks bb2 WHERE SUBSTR(bb2.header, 5, 32) = bb1.hash)
+		
+		AND NOT EXISTS (
+			SELECT * FROM btc_blocks bb3 WHERE bb1.height = bb3.height 
+			AND EXISTS(
+				SELECT * FROM btc_blocks bb4 WHERE substr(bb4.header, 5, 32) = bb3.hash
+			)
+		)
+		ORDER BY height DESC
+		OFFSET $1 + 1
+		LIMIT 100
```

Comment:
Is there anyway we could further optimise this query? It appears to select from `btc_blocks` four times.

Also there's a few inconsistencies that I think we could fix:
```suggestion
		SELECT height FROM btc_blocks bb1
		WHERE NOT EXISTS (SELECT * FROM btc_blocks bb2 WHERE substr(bb2.header, 5, 32) = bb1.hash)
		AND NOT EXISTS (
			SELECT * FROM btc_blocks bb3 WHERE bb1.height = bb3.height 
			AND EXISTS (
				SELECT * FROM btc_blocks bb4 WHERE substr(bb4.header, 5, 32) = bb3.hash
			)
		)
		ORDER BY height DESC
		OFFSET $1 + 1
		LIMIT 100
```

Reply:
hey @joshuasing ok so I did some digging into whether we can optimize this.

I tested this locally, so I ran `EXPLAIN` with both `set enable_seqscan = off` and `set enable_seqscan = on`.

With that `on`, meaning we don't discourage the use of seq scans, the planner doesn't help much, as I have so few rows in my local db that it determines the best way to query is to use seq scans.
```sql
                                                 QUERY PLAN                                                 
------------------------------------------------------------------------------------------------------------
 Limit  (cost=112.47..112.72 rows=100 width=8)
   ->  Sort  (cost=112.47..112.90 rows=175 width=8)
         Sort Key: bb1.height DESC
         ->  Hash Anti Join  (cost=76.59..105.95 rows=175 width=8)
               Hash Cond: (bb1.height = bb3.height)
               ->  Hash Right Anti Join  (cost=25.75..48.96 rows=350 width=8)
                     Hash Cond: (substr(bb2.header, 5, 32) = bb1.hash)
                     ->  Seq Scan on btc_blocks bb2  (cost=0.00..17.00 rows=700 width=32)
                     ->  Hash  (cost=17.00..17.00 rows=700 width=40)
                           ->  Seq Scan on btc_blocks bb1  (cost=0.00..17.00 rows=700 width=40)
               ->  Hash  (cost=46.46..46.46 rows=350 width=8)
                     ->  Hash Join  (cost=23.25..46.46 rows=350 width=8)
                           Hash Cond: (bb3.hash = substr(bb4.header, 5, 32))
                           ->  Seq Scan on btc_blocks bb3  (cost=0.00..17.00 rows=700 width=40)
                           ->  Hash  (cost=20.75..20.75 rows=200 width=32)
                                 ->  HashAggregate  (cost=18.75..20.75 rows=200 width=32)
                                       Group Key: substr(bb4.header, 5, 32)
                                       ->  Seq Scan on btc_blocks bb4  (cost=0.00..17.00 rows=700 width=32)
(18 rows)
```

so, if I _discourage_ the use of seq scans, I can see postgres uses the following indexes:
```sql
                                                         QUERY PLAN                                                         
----------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=203.42..203.67 rows=100 width=8)
   ->  Sort  (cost=203.42..203.85 rows=175 width=8)
         Sort Key: bb1.height DESC
         ->  Hash Anti Join  (cost=167.54..196.90 rows=175 width=8)
               Hash Cond: (bb1.height = bb3.height)
               ->  Hash Right Anti Join  (cost=71.22..94.44 rows=350 width=8)
                     Hash Cond: (substr(bb2.header, 5, 32) = bb1.hash)
                     ->  Bitmap Heap Scan on btc_blocks bb2  (cost=7.83..24.82 rows=700 width=32)
                           ->  Bitmap Index Scan on btc_blocks_hash  (cost=0.00..7.65 rows=700 width=0)
                     ->  Hash  (cost=54.65..54.65 rows=700 width=40)
                           ->  Index Scan using height_idx on btc_blocks bb1  (cost=0.15..54.65 rows=700 width=40)
               ->  Hash  (cost=91.94..91.94 rows=350 width=8)
                     ->  Hash Join  (cost=31.22..91.94 rows=350 width=8)
                           Hash Cond: (bb3.hash = substr(bb4.header, 5, 32))
                           ->  Index Scan Backward using height_idx on btc_blocks bb3  (cost=0.15..54.65 rows=700 width=40)
                           ->  Hash  (cost=28.57..28.57 rows=200 width=32)
                                 ->  HashAggregate  (cost=26.57..28.57 rows=200 width=32)
                                       Group Key: substr(bb4.header, 5, 32)
                                       ->  Bitmap Heap Scan on btc_blocks bb4  (cost=7.83..24.82 rows=700 width=32)
                                             ->  Bitmap Index Scan on btc_blocks_hash  (cost=0.00..7.65 rows=700 width=0)
(20 rows)
```

whilst we do `SELECT FROM btc_blocks` quite a bit, the query plan is given the _option_ to use the indexes  `height_idx` and `btc_blocks_hash` to optimize the query.


so, given there are no sql changes, I can run this on testnet to see what we would do there:
```sql
                                                                                       QUERY PLAN                                                                                        
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=76187.25..76198.91 rows=100 width=8) (actual time=508.866..520.173 rows=74 loops=1)
   ->  Gather Merge  (cost=76187.13..82027.40 rows=50056 width=8) (actual time=508.863..520.165 rows=75 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=75187.11..75249.68 rows=25028 width=8) (actual time=505.763..505.771 rows=25 loops=3)
               Sort Key: bb1.height DESC
               Sort Method: quicksort  Memory: 25kB
               Worker 0:  Sort Method: quicksort  Memory: 25kB
               Worker 1:  Sort Method: quicksort  Memory: 25kB
               ->  Parallel Hash Right Anti Join  (cost=67606.48..74228.76 rows=25028 width=8) (actual time=505.535..505.713 rows=25 loops=3)
                     Hash Cond: (substr(bb2.header, 5, 32) = bb1.hash)
                     ->  Parallel Seq Scan on btc_blocks bb2  (cost=0.00..5913.14 rows=100114 width=81) (actual time=0.005..6.826 rows=80091 loops=3)
                     ->  Parallel Hash  (cost=66980.77..66980.77 rows=50057 width=41) (actual time=484.837..484.843 rows=25 loops=3)
                           Buckets: 131072  Batches: 1  Memory Usage: 1120kB
                           ->  Parallel Hash Anti Join  (cost=60247.95..66980.77 rows=50057 width=41) (actual time=450.641..484.612 rows=25 loops=3)
                                 Hash Cond: (bb1.height = bb3.height)
                                 ->  Parallel Seq Scan on btc_blocks bb1  (cost=0.00..5913.14 rows=100114 width=41) (actual time=0.011..8.728 rows=80091 loops=3)
                                 ->  Parallel Hash  (cost=59622.23..59622.23 rows=50057 width=8) (actual time=450.053..450.057 rows=80064 loops=3)
                                       Buckets: 262144 (originally 131072)  Batches: 1 (originally 1)  Memory Usage: 12544kB
                                       ->  Hash Join  (cost=47775.58..59622.23 rows=50057 width=8) (actual time=308.685..395.255 rows=80064 loops=3)
                                             Hash Cond: (bb3.hash = (substr(bb4.header, 5, 32)))
                                             ->  Parallel Seq Scan on btc_blocks bb3  (cost=0.00..5913.14 rows=100114 width=41) (actual time=0.011..9.082 rows=80091 loops=3)
                                             ->  Hash  (cost=41487.17..41487.17 rows=240273 width=81) (actual time=308.364..308.367 rows=240269 loops=3)
                                                   Buckets: 65536 (originally 65536)  Batches: 8 (originally 4)  Memory Usage: 7681kB
                                                   ->  Unique  (cost=40285.81..41487.17 rows=240273 width=81) (actual time=153.986..244.353 rows=240269 loops=3)
                                                         ->  Sort  (cost=40285.81..40886.49 rows=240273 width=81) (actual time=153.985..207.116 rows=240273 loops=3)
                                                               Sort Key: (substr(bb4.header, 5, 32))
                                                               Sort Method: external merge  Disk: 29176kB
                                                               Worker 0:  Sort Method: external merge  Disk: 29176kB
                                                               Worker 1:  Sort Method: external merge  Disk: 29176kB
                                                               ->  Seq Scan on btc_blocks bb4  (cost=0.00..7314.73 rows=240273 width=81) (actual time=0.016..47.489 rows=240273 loops=3)
 Planning Time: 0.397 ms
 Execution Time: 526.158 ms
(33 rows)
```

that's not good; that cost is too high, so if I try to add and functional index on testnet (don't worry I'll remove it)

```sql
CREATE INDEX btc_blocks_header_prev_hash_idx ON btc_blocks (substr(header, 5, 32));
```

```sql
                                                                                    QUERY PLAN                                                                                    
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1051.40..6021.14 rows=100 width=8) (actual time=321.047..325.646 rows=74 loops=1)
   ->  Nested Loop Anti Join  (cost=1001.70..2986224.09 rows=60068 width=8) (actual time=319.796..325.638 rows=75 loops=1)
         ->  Gather Merge  (cost=1000.86..1175528.49 rows=120136 width=8) (actual time=319.736..325.059 rows=80 loops=1)
               Workers Planned: 2
               Workers Launched: 2
               ->  Nested Loop Anti Join  (cost=0.84..1160661.79 rows=50057 width=8) (actual time=1.452..211.113 rows=27 loops=3)
                     ->  Parallel Index Scan using height_idx on btc_blocks bb1  (cost=0.42..11868.81 rows=100114 width=41) (actual time=0.013..24.873 rows=80091 loops=3)
                     ->  Index Scan using btc_blocks_header_prev_hash_idx on btc_blocks bb2  (cost=0.42..21.55 rows=1201 width=81) (actual time=0.002..0.002 rows=1 loops=240273)
                           Index Cond: (substr(header, 5, 32) = bb1.hash)
         ->  Nested Loop Semi Join  (cost=0.84..15.06 rows=1 width=8) (actual time=0.007..0.007 rows=0 loops=80)
               ->  Index Scan using height_idx on btc_blocks bb3  (cost=0.42..0.54 rows=1 width=41) (actual time=0.003..0.003 rows=1 loops=80)
                     Index Cond: (height = bb1.height)
               ->  Index Scan using btc_blocks_header_prev_hash_idx on btc_blocks bb4  (cost=0.42..21.55 rows=1201 width=81) (actual time=0.003..0.003 rows=0 loops=84)
                     Index Cond: (substr(header, 5, 32) = bb3.hash)
 Planning Time: 0.590 ms
 Execution Time: 325.682 ms
(16 rows)
```

I get 1/75 of the cost! wow! ok so I am going to add that index to the pr, good call 👍 🤯